pkg/segment/segment_test.go

```diff
@@ -0,0 +1,191 @@
+package segment
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/preference"
+	"github.com/openshift/odo/pkg/version"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+type segmentResponse struct {
+	Batch []struct {
+		AnonymousID string `json:"anonymousId"`
+		MessageId   string `json:"messageId"`
+		Traits      struct {
+			OS string `json:"os"`
+		} `json:"traits"`
+		Properties struct {
+			Error     string `json:"error"`
+			ErrorType string `json:"error-type"`
+			Success   bool   `json:"success"`
+			Version   string `json:"version"`
+		} `json:"properties"`
+		Type string `json:"type"`
+	} `json:"batch"`
+	MessageID string `json:"messageId"`
+}
+
+func mockServer() (chan []byte, *httptest.Server) {
+	done := make(chan []byte, 1)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		bin, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		done <- bin
+	}))
+	return done, server
+}
+
+func TestClientUploadWithoutConsent(t *testing.T) {
+	body, server := mockServer()
+	defer server.Close()
+	defer close(body)
+	falseValue := false
+
+	cfg := &preference.PreferenceInfo{
+		Preference: preference.Preference{
+			OdoSettings: preference.OdoSettings{
+				ConsentTelemetry: &falseValue,
+			},
+		},
+	}
+
+	c, err := newCustomClient(cfg, filepath.Join(os.TempDir(), "telemetry"), server.URL)
+	//	assert err != nil
+	if err != nil {
+		t.Fatal(err)
+	}
+	// run a command, odo preference view
+	if err := c.Upload("odo preference view", time.Second, errors.New("an error occured")); err != nil {
+		t.Fatal(err)
+	}
+	if err := c.Close(); err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case <-body:
+		t.Fatal("server should not receive data")
+	default:
+	}
+}
+
+func TestClientUploadWithConsent(t *testing.T) {
+	body, server := mockServer()
+	defer server.Close()
+	defer close(body)
+
+	trueValue := true
+
+	cfg := &preference.PreferenceInfo{
+		Preference: preference.Preference{
+			OdoSettings: preference.OdoSettings{
+				ConsentTelemetry: &trueValue,
+			},
+		},
+	}
+	tests := []struct {
+		testName string
+		err      error
+		success  bool
+		errType  string
+	}{
+		{
+			testName: "command ran successfully",
+			err:      nil,
+			success:  true,
+			errType:  "",
+		},
+		{
+			testName: "command failed",
+			err:      errors.New("some error occurred"),
+			success:  false,
+			errType:  "*errors.errorString",
+		},
+	}
+	for _, tt := range tests {
+		t.Log("Running test: ", tt.testName)
+		t.Run(tt.testName, func(t *testing.T) {
+			c, err := newCustomClient(cfg, filepath.Join(os.TempDir(), "telemetry"), server.URL)
```

Comment:
A unit test shouldn't interact with external system that are not now controlled by you e.g. filesystems, API etc..
Please mock the filesystem using afero - you will find many such examples